### PR TITLE
Explore: Consolidate logs volume logic (full range and limited)

### LINKED ISSUE #61994: Explore: Consolidate logic to create full-range or results-based histogram
### Background

Current logic for logs volume is implemented in two ways:
* (1) Data source may return a provider
* (2) Explore collects all returned logs and creates a histogram based on results

To make implementing mixed data sources easier and simplify the logic in general, (2) can be abstracted in a way that it becomes a data provider that Explore provides to data sources without full-range support.

This should also simplify rendering that is currently conditional based on wether data source supports Logs Volume data provider. If we always rely on logic based on providers the view should not need any conditional checks.

### Requirements ###

High level: _Ensure full-range and result-based logs volume is created through the same flow and stored as supplementary query result_

#### Technical details

* Explore returns result-based logs only of logs volume provider is not available [here](https://github.com/grafana/grafana/blob/main/public/app/features/explore/utils/decorators.ts#L239). This check can be removed as we will always have a provider (full-range from the data source or result-based fallback).
* Explore needs to create results-based logs provider that will be based on results. Hence the provider will need access to the [query subscription](https://github.com/grafana/grafana/blob/main/public/app/features/explore/state/query.ts#L467) (with and without caching).
* The logic of the results-based log volume can simply use returned ExplorePanelData.logsResult.series (LogsModel), just wrap passing it as an Observable<DataQueryResponse> to conform with supplementary queries API
* Any new added types should be marked as `@internal`

<details><summary>DoR checklist</summary>

- [x] Should our squad work on it?  ***Yes, part of mixed data sources initiative***
- [x] Why is it important?  ***This task doesn't provide new functionalities but simplifies the code and reduces tech debt***
- [x] Dependencies: ***No***
- [x] Priority: ***N/A***
- [x] Functional requirements. ***No changes in behaviour***
- [x] Non-functional requirements . ***N/A***
- [X] Documentation requirements. ***Only changed in API, but there should be no breaking changes***
- [x] UX design / approval. ***Not needed***
- [x] Technical design / approval.  ***See technical details***
- [x] How will the outcome value of the feature be measured? ***N/A***
- [x] Should we write a blog post about it?  ***No***
</details>

### PR DESCRIPTION
**Pull request by ifrost**
This is to simplify logic around creating Logs Volume. There should be no changes in the current behavior.

How to test it?
Ensure that creating a full range and results-based logs volume (mixed/non-mixed) produces the same results in this branch and in main.

<details><summary>example</summary>

| Example | Screenshot |
|--|--|
| TestDataSource (full range logs volume not supported | Screenshot 2023-01-30 at 10 09 14 |
| Loki (full range logs volume supported |  Screenshot 2023-01-30 at 10 14 16 |
| Mixed (always shows limited logs volume |  Screenshot 2023-01-30 at 10 14 51|

</details>

Fixes #61994

### CODE REVIEW

File: public/app/core/logsModel.test.ts [hunk 1]
```diff
@@ -1186,6 +1187,43 @@ describe('logs volume', () => {
...
+                    },
+                    {
+                      target: 'volume query 2',
+                    },
+                  ],
+                  logsVolumeType: LogsVolumeType.FullRange,
+                  absoluteRange: {
+                    from: 1623888000000,
+                    to: 1623888000000,
+                  },
```
Comment (Elfo404) [inline public/app/core/logsModel.test.ts:1216]: super tiny nit: i had a bit of a hard time understanding this and realizing it is hardcoded `2021-06-17 00:00:00`. could we somehow calculate this from whatever we use in the setup function?

File: public/app/features/explore/state/query.ts [hunk 2]
```diff
@@ -447,7 +447,8 @@ export const runQueries = (
       cache,
       supplementaryQueries,
     } = exploreItemState;
-    let newQuerySub;
+    let newQuerySource: Observable<ExplorePanelData>;
+    let newQuerySubscription: SubscriptionLike | undefined;
```
Comment (Elfo404) [inline public/app/features/explore/state/query.ts:451]: we never try to use it before assignment, so 
```suggestion
    let newQuerySubscription: SubscriptionLike;
```
works as well

**Review by Elfo404 — APPROVED** [review]
look and works great! just a couple of nits.

side note: i'm changing the milestone of this to 9.5

File: public/app/core/logsModel.test.ts (hunk 1 above)
Comment (ifrost) [reply public/app/core/logsModel.test.ts]: Updated in 1d604dfc1e8d3f75d07059c29dd17e6b585c9a6e

File: public/app/features/explore/state/query.ts (hunk 2 above)
Comment (ifrost) [reply public/app/features/explore/state/query.ts]: Updated in f4190d1243f4431dabfa091ed71551ac95eb1b55

File: public/app/features/explore/state/query.ts [hunk 3]
```diff
@@ -593,13 +574,19 @@ export const runQueries = (
...
+            {
               ...transaction.request,
               requestId: `${transaction.request.requestId}_${snakeCase(type)}`,
-            });
+            },
+            newQuerySource
+          );
+
+          // We always prepare provider, even is supplementary query is disabled because when the user
+          // enables the query, we need to load the data, so we need the provider
```
Comment (ivanahuckova) [inline public/app/features/explore/state/query.ts:588]: Is this the right place for the comment? Should't it be above `const dataProvider =...`

File: public/app/features/explore/LogsVolumePanel.tsx [hunk 4]
```diff
@@ -29,56 +30,20 @@ type Props = {
...
-
-  if (data === undefined) {
+  if (props.logsVolumeData === undefined) {
     return null;
   }
 
-  const { logsVolumeData, fullRangeData, range } = data;
+  const logsVolumeData = props.logsVolumeData;
+  const range = logsVolumeData.data[0]?.meta?.custom?.absoluteRange || props.absoluteRange;
+  const fullRangeData = logsVolumeData.data[0]?.meta?.custom?.logsVolumeType !== LogsVolumeType.Limited;
```
Comment (ivanahuckova) [inline public/app/features/explore/LogsVolumePanel.tsx:46]: I would either rename `fullRangeData` to `isFullRangeData` as the value is boolean. Or (in my opinion better), do `const logsVolumeType = logsVolumeData.data[0]?.meta?.custom?.logsVolumeType` and use that below.

**Review by ivanahuckova — COMMENTED** [review]
Nice work with this! 🤩

File: public/app/core/logsModel.test.ts [hunk 5]
```diff
@@ -31,6 +32,9 @@ import {
   queryLogsSample,
 } from './logsModel';
 
+const FROM = dateTimeParse('2021-06-17 00:00:00', { timeZone: 'utc' });
+const TO = dateTimeParse('2021-06-17 00:00:00', { timeZone: 'utc' });
```
Comment (Elfo404) [inline public/app/core/logsModel.test.ts:36]: ❤️

File: public/app/features/explore/state/query.ts (hunk 3 above)
Comment (ifrost) [reply public/app/features/explore/state/query.ts]: Fixed in 7da84bd360ccdf3eec5605ed4af27a2c03a73621

File: public/app/features/explore/LogsVolumePanel.tsx (hunk 4 above)
Comment (ifrost) [reply public/app/features/explore/LogsVolumePanel.tsx:46]: Good point! Fixed in 7ce66774d2af6dc6444f37fe358261a7b321592a

**Review by ivanahuckova — APPROVED** [review]
Yay! LGTM.

**Pull request merged**